rts/lra/lra-proxy/api/src/main/java/io/narayana/lra/client/internal/proxy/nonjaxrs/LRAParticipant.java

```diff
@@ -0,0 +1,249 @@
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package io.narayana.lra.client.internal.proxy.nonjaxrs;
+
+import org.eclipse.microprofile.lra.annotation.ParticipantStatus;
+import org.eclipse.microprofile.lra.participant.InvalidLRAParticipantDefinitionException;
+
+import javax.enterprise.inject.spi.CDI;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.UriBuilder;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.net.URI;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletionStage;
+
+import static io.narayana.lra.LRAConstants.COMPENSATE;
+import static io.narayana.lra.LRAConstants.COMPLETE;
+import static io.narayana.lra.LRAConstants.FORGET;
+import static io.narayana.lra.LRAConstants.STATUS;
+
+/**
+ * Keeps references to individual non-JAX-RS paraticipant methods in
+ * single LRA participant class.
+ */
+public class LRAParticipant {
+
+    private Class<?> id;
+    private Method compensateMethod;
+    private Method completeMethod;
+    private Method statusMethod;
+    private Method forgetMethod;
+
+    private Map<URI, Optional<?>> participantStatusMap = new HashMap<>();
+
+    LRAParticipant(Class<?> id) {
+        this.id = id;
+    }
+
+    Class<?> getId() {
+        return id;
+    }
+
+    Response compensate(URI lraId, URI parentId) {
+        return invokeParticipantMethod(compensateMethod, lraId, parentId, COMPENSATE);
+    }
+
+    Response complete(URI lraId, URI parentId) {
+        return invokeParticipantMethod(completeMethod, lraId, parentId, COMPLETE);
+    }
+
+    Response status(URI lraId, URI parentId) {
+        if (participantStatusMap.containsKey(lraId)) {
+            Optional<?> optional = participantStatusMap.get(lraId);
+            if (optional.isPresent()) {
+                participantStatusMap.remove(lraId);
+                
+                Object result = optional.get();
+                if (shouldInvokeStatusMethod(result)) {
+                    return invokeParticipantMethod(statusMethod, lraId, parentId, STATUS);
```

Comment:
I am relying here on coordinator because it is calling status on the participant right now it must have found out that this participant has a status endpoint thus method. However, this a good catch because in case there is no status compensate or complete will get reinvoked, so we should apply same logic to both of this methods. I will update the PR later today. Thanks!